src/server.rs

```diff
@@ -519,3 +675,148 @@ impl Drop for ServerManager {
+
+    unsafe fn from_raw(raw: *mut c_void) -> Self {
+        let val = &mut *(raw as *mut sys::GSClientApprove_t);
+        GSClientApprove {
+            user: SteamId(val.m_SteamID.m_steamid.m_unAll64Bits),
+            owner: SteamId(val.m_OwnerSteamID.m_steamid.m_unAll64Bits),
+        }
+    }
+}
+
+/// Used to set the mode that a gameserver will run in
+#[repr(i32)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum DenyReason {
+    Invalid,
+    InvalidVersion,
+    Generic,
+    NotLoggedOn,
+    NoLicense,
+    Cheater,
+    LoggedInElseWhere,
+    UnknownText,
+    IncompatibleAnticheat,
+    MemoryCorruption,
+    IncompatibleSoftware,
+    SteamConnectionLost,
+    SteamConnectionError,
+    SteamResponseTimedOut,
+    SteamValidationStalled,
+    SteamOwnerLeftGuestUser,
+}
+
+impl From<sys::EDenyReason> for DenyReason {
+    fn from(r: sys::EDenyReason) -> Self {
+        match r {
+            sys::EDenyReason::k_EDenyInvalid => DenyReason::Invalid,
+            sys::EDenyReason::k_EDenyInvalidVersion => DenyReason::InvalidVersion,
+            sys::EDenyReason::k_EDenyGeneric => DenyReason::Generic,
+            sys::EDenyReason::k_EDenyNotLoggedOn => DenyReason::NotLoggedOn,
+            sys::EDenyReason::k_EDenyNoLicense => DenyReason::NoLicense,
+            sys::EDenyReason::k_EDenyCheater => DenyReason::Cheater,
+            sys::EDenyReason::k_EDenyLoggedInElseWhere => DenyReason::LoggedInElseWhere,
+            sys::EDenyReason::k_EDenyUnknownText => DenyReason::UnknownText,
+            sys::EDenyReason::k_EDenyIncompatibleAnticheat => DenyReason::IncompatibleAnticheat,
+            sys::EDenyReason::k_EDenyMemoryCorruption => DenyReason::MemoryCorruption,
+            sys::EDenyReason::k_EDenyIncompatibleSoftware => DenyReason::IncompatibleSoftware,
+            sys::EDenyReason::k_EDenySteamConnectionLost => DenyReason::SteamConnectionLost,
+            sys::EDenyReason::k_EDenySteamConnectionError => DenyReason::SteamConnectionError,
+            sys::EDenyReason::k_EDenySteamResponseTimedOut => DenyReason::SteamResponseTimedOut,
+            sys::EDenyReason::k_EDenySteamValidationStalled => DenyReason::SteamValidationStalled,
+            sys::EDenyReason::k_EDenySteamOwnerLeftGuestUser => DenyReason::SteamOwnerLeftGuestUser,
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// Called when a user has been denied to connection to this game server
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct GSClientDeny {
+    /// The steam ID of the user requesting a p2p session
+    pub user: SteamId,
+    pub deny_reason: DenyReason,
+    pub optional_text: String,
+}
+
+unsafe impl Callback for GSClientDeny {
+    const ID: i32 = steamworks_sys::GSClientDeny_t_k_iCallback as i32;
+
+    unsafe fn from_raw(raw: *mut c_void) -> Self {
+        let val = &mut *(raw as *mut sys::GSClientDeny_t);
```

Comment:
To match up with #260, use ptr::read_unaligned here instead of casting to a mutable borrow, since we don't know if the provided pointers are properly aligned or not.